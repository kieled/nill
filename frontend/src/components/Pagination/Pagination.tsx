'use client'

import {FC, useMemo} from "react"
import cn from "classnames"

type Props = {
  pageCount: number;
  visiblePageButtonsCount: number;
  currentPage: number;
  onPageChange: (pageNumber: number) => void;
}

const Pagination: FC<Props> = (
  {
    pageCount,
    visiblePageButtonsCount,
    currentPage,
    onPageChange,
  }
) => {
  if (pageCount === 1) {
    return <></>
  }

  const visiblePages = useMemo(() => {
    const startPage =
      currentPage <= Math.floor(visiblePageButtonsCount / 2)
        ? 1
        : Math.min(
          currentPage - Math.floor(visiblePageButtonsCount / 2),
          pageCount - visiblePageButtonsCount + 1
        )
    const endPage = Math.min(startPage + visiblePageButtonsCount - 1, pageCount)
    return Array.from(
      {length: endPage - startPage + 1},
      (_, i) => startPage + i
    )
  }, [currentPage, pageCount, visiblePageButtonsCount])

  const handlePageChange = (pageNumber: number) => {
    if (pageNumber < 1 || pageNumber > pageCount) {
      return
    }

    onPageChange(pageNumber)
  }

  const handlePrevPageClick = () => {
    if (currentPage > 1) {
      handlePageChange(currentPage - 1)
    }
  }

  const handleNextPageClick = () => {
    if (currentPage < pageCount) {
      handlePageChange(currentPage + 1)
    }
  }

  return (
    <div>
      <button data-testid='prev' onClick={handlePrevPageClick} disabled={currentPage === 1}>
        {"<"}
      </button>
      {visiblePages.map((pageNumber) => (
        <button
          key={pageNumber}
          onClick={() => handlePageChange(pageNumber)}
          className={cn({'active': currentPage === pageNumber})}
        >
          {pageNumber}
        </button>
      ))}
      <button data-testid='next' onClick={handleNextPageClick} disabled={currentPage === pageCount}>
        {">"}
      </button>
    </div>
  )
}

export default Pagination
