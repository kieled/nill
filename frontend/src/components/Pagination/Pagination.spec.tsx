import React from "react"
import {fireEvent, render} from "@testing-library/react"
import Pagination from "./Pagination"

describe("Pagination component", () => {
  const defaultProps = {
    pageCount: 10,
    visiblePageButtonsCount: 5,
    currentPage: 1,
    onPageChange: jest.fn(),
  }

  afterEach(() => {
    jest.clearAllMocks()
  })

  it("renders correctly", () => {
    const {container} = render(<Pagination {...defaultProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders no buttons when there is only one page", () => {
    const {container} = render(<Pagination {...defaultProps} pageCount={1}/>)
    expect(container.firstChild).toBeNull()
  })

  it("calculates the visible pages correctly", () => {
    const {getAllByRole} = render(<Pagination {...defaultProps} />)
    const pageButtons = getAllByRole("button").slice(1, -1)
    const pageNumbers = pageButtons.map((button) => Number(button.textContent))
    expect(pageNumbers).toEqual([1, 2, 3, 4, 5])
  })

  it("calculates the visible pages correctly when currentPage is in the middle", () => {
    const {getAllByRole} = render(<Pagination {...defaultProps} currentPage={5}/>)
    const pageButtons = getAllByRole("button").slice(1, -1)
    const pageNumbers = pageButtons.map((button) => Number(button.textContent))
    expect(pageNumbers).toEqual([3, 4, 5, 6, 7])
  })

  it("calculates the visible pages correctly when currentPage is at the end", () => {
    const {getAllByRole} = render(<Pagination {...defaultProps} currentPage={10}/>)
    const pageButtons = getAllByRole("button").slice(1, -1)
    const pageNumbers = pageButtons.map((button) => Number(button.textContent))
    expect(pageNumbers).toEqual([6, 7, 8, 9, 10])
  })

  it("calls onPageChange with the correct page number when a page button is clicked", () => {
    const {getAllByRole} = render(<Pagination {...defaultProps} />)
    const pageButtons = getAllByRole("button").slice(1, -1)
    const page3Button = pageButtons.find((button) => button.textContent === "3")
    fireEvent.click(page3Button as HTMLElement)
    expect(defaultProps.onPageChange).toHaveBeenCalledWith(3)
  })

  it("calls onPageChange with the correct page number when the previous page button is clicked", () => {
    const {getByTestId} = render(<Pagination {...defaultProps} currentPage={3}/>)
    fireEvent.click(getByTestId('prev'))
    expect(defaultProps.onPageChange).toHaveBeenCalledWith(2)
  })

  it("calls onPageChange with the correct page number when the next page button is clicked", () => {
    const {getByTestId} = render(<Pagination {...defaultProps} currentPage={3}/>)
    fireEvent.click(getByTestId('next'))
    expect(defaultProps.onPageChange).toHaveBeenCalledWith(4)
  })

  it("disables the previous page button when currentPage is 1", () => {
    const {getAllByRole} = render(<Pagination {...defaultProps} currentPage={1}/>)
    const [prevPageButton] = getAllByRole("button")
    expect(prevPageButton).toBeDisabled()
  })
})