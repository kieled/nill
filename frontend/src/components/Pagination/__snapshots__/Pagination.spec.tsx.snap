// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component renders correctly 1`] = `
<div>
  <button
    data-testid="prev"
    disabled=""
  >
    &lt;
  </button>
  <button
    class="active"
  >
    1
  </button>
  <button
    class=""
  >
    2
  </button>
  <button
    class=""
  >
    3
  </button>
  <button
    class=""
  >
    4
  </button>
  <button
    class=""
  >
    5
  </button>
  <button
    data-testid="next"
  >
    &gt;
  </button>
</div>
`;
